lib/matplotlib/backend_bases.py

```diff
@@ -1394,14 +1394,16 @@ def __init__(self, name, canvas, x, y, button=None, key=None,
         (*x*, *y*) in figure coords ((0, 0) = bottom left)
         button pressed None, 1, 2, 3, 'up', 'down'
         """
-        LocationEvent.__init__(self, name, canvas, x, y, guiEvent=guiEvent)
         if button in MouseButton.__members__.values():
             button = MouseButton(button)
         self.button = button
         self.key = key
         self.step = step
         self.dblclick = dblclick
 
+        # super-init deferred to the end: callback errors if called before
```

Comment:
Please include the reason why this is necessary (not just stating that it doesn't work otherwise :smile:):
```suggestion
        # super-init is deferred to the end because it calls back on
        # 'axes_enter_event', which requires a fully initialized event.
```

Given this explanation, I'm not 100% sure that this change is the correct solution. `__init__` calling other methods that rely on a fully set-up object can be problematic (as this issue illustrates).

More generally, the mechanism that the creation of one event triggers callbacks for another event looks like we have not separated responsibilities well enough. IMHO the creator of the `LocationEvent` (or its subclasses) should make sure that the enter/leave state is updated.

Currently we have the subclasses `MouseEvent` and `KeyEvent`. IMHO only the 'motion_notify_event' subtype of `MouseEvent` is relevant. Neither mouse presses nor key presses should modify the enter/leave state. The only place we create the 'motion_notify_event' subtype of `MouseEvent` is in `FigureCanvasBase.motion_notify_event`. Could this explicitly check the enter/leave state there and process the callbacks directly?

This may be related to or interfere with #16931.

Reply:
An Alt+Tab is a key press that could alter enter/leave state, no?

Reply:
I haven't checked, but would we get any event if the window is made active and the cursor happens to be on an axes? E.g. you close another window and focus is given to the plot window and by chance the cursor is in there.

My naive solution for enter/leave event would be to only check mouse movements and detect if the axes under the cursor has changed. That would not include cases in which the mouse is brought to an axes by means other than mouse movements, but as soon as you start moving the mouse, the new axes would be detected. IMHO that's simple and good enough.

Reply:
I checked, and you don't get leave if you <kbd>Alt+Tab</kbd> out, but you get enter if you <kbd>Alt+Tab</kbd> in.
```
fig, ax = plt.subplots()
for ev in ['axes_enter', 'axes_leave', 'key_press', 'key_release', 'mouse_press', 'mouse_release']:
    fig.canvas.mpl_connect(f'{ev}_event', lambda event, ev=ev: print(ev, repr(event)))
```
(Note you need `repr` until this PR is merged.)